rubocop: document.rb, convertible.rb and renderer.rb

ayastreb opened the pull request:
#4885 

Ok, now I think I'm finished with this 😄 
Main goal of this PR is to remove code duplication in `document.rb`, `renderer.rb` and `convertible.rb` and make them all Rubocop compliant.

I included `Convertible` into `Document` and implemented `render` method the same way it is implemented in `Page`
I removed all the logic from `Renderer` class and just call this `render` method on the document instance.

I did some refactoring in `convertible.rb`, mainly:
- move private methods to the bottom of the class
- move payload from methods argument to class attribute

Feedback is welcome! 😄

envygeeks commented on lib/jekyll/convertible.rb:
```diff
@@ -34,58 +36,17 @@ def published?
     # opts - optional parameter to File.read, default at site configs
     #
-    # Returns nothing.
+    # Returns data read from file.
```
`Returns [Type] description`

ayastreb commented:
I think it would be better to include Convertible in Document, and then just call render on document instance in Renderer.
I will try to do that and merge this PR with another one for Document

envygeeks commented:
hey @ayastreb can you clean up the commit history so it's easier to read? Thanks!

ayastreb commented:
Hi @envygeeks I've squashed the commits and rebased, but it looks like that same `test_site.rb` failed again 😄 
Could you please re-start the tests?

envygeeks commented:
Hulk here,

I've restarted the failing tests for you.
Thanks.

parkr commented:
> I included Convertible into Document and implemented render method the same way it is implemented in Page
> I removed all the logic from Renderer class and just call this render method on the document instance.

@ayastreb This is the opposite of our intent on creating the `Render` class. The idea was that we'd use a new instance of the `Render` class to render documents, and any document passed in would have to respond to a given interface.

Ultimately, we want to _remove_ the `Convertible` module in favor of the `Render` class. So much of your refactoring here is not acceptable in its current form. Would you please rework it?

ayastreb commented:
@parkr, I'we reworked this PR and moved back the rendering logic to `Renderer` class.
I removed all the rendering logic from `Convertible` class and now it seems to be doing other things: reading yaml files, writing files and checking file types. Some methods (`published?`, `write`, `sass_file?`, etc) are still duplicated in `Document` class. Maybe we should split or rename `Convertible` somehow?

In `Page` class I've changed `render` method to use `Renderer` instead of using methods from `Convertible`. I also changed `Page.render` signature - we don't need `layouts` variable there.
Do you think it's ok? Could it be that other people use this method somehow? I only found this method used in tests, so not sure 😕

ayastreb commented:
Hi @parkr, just pinging here 😄 
Is there anything I could do more in this PR?

parkr commented:
@ayastreb This PR now has a number of merge conflicts. Would you mind rebasing on the latest `origin/master`?

ayastreb commented:
Hi @parkr, I rebased my branch but looks like that test failed again.
It fails from time to time randomly :(
Could you restar the build, please?

parkr commented on lib/jekyll/convertible.rb:
```diff
@@ -34,58 +37,17 @@ def published?
-
-      self.data ||= {}
-
+      read_data filename, opts
```
Let's call this method `read_and_parse_contents` rather than `read_data`.

parkr commented on lib/jekyll/document.rb:
```diff
@@ -7,9 +7,9 @@ class Document
-    DATE_FILENAME_MATCHER = /^(.+\/)*(\d+-\d+-\d+)-(.*)(\.[^.]+)$/
+    YAML_FRONT_MATTER_REGEXP = %r!\A(---\s*\n.*?\n?)^((---|\.\.\.)\s*$\n?)!m
+    DATELESS_FILENAME_MATCHER = %r!^(?:.+\/)*(.*)(\.[^.]+)$!
+    DATE_FILENAME_MATCHER = %r!^(?:.+\/)*(\d+-\d+-\d+)-(.*)(\.[^.]+)$!
```
Why did you add the `?:`? That doesn't seem necessary? You trying to pull a fast one on me? 😂

ayastreb replied:
I added it to make the first group not captured, so that I can use all the captures later like this:
```ruby
if relative_path =~ DATE_FILENAME_MATCHER
  date, slug, ext = Regexp.last_match.captures
```
instead of
```ruby
if relative_path =~ DATE_FILENAME_MATCHER
   date, slug, ext = $2, $3, $4
```
So I changed the regexp to only capture date, slug and ext so that I don't need to use $2, $3, $4 (Rubocop doesn't like Pearl-style captures)

parkr commented:
@ayastreb For the record, there are definitely some substantive changes in here! Code clean-up is easy to review but larger changes, e.g. switching Pathname for Pathutil, need to be done very carefully and in an isolated manner so we don't break anything.

parkr commented on lib/jekyll/convertible.rb:
```diff
@@ -278,10 +111,8 @@ def do_layout(payload, layouts)
-        f.write(output)
-      end
-      Jekyll::Hooks.trigger hook_owner, :post_write, self
+      File.write(path, output)
```
The `wb` mode helped with content that was non-ASCII. We need to make sure we keep that mode.

parkr commented on lib/jekyll/convertible.rb:
```diff
@@ -296,5 +127,42 @@ def [](property)
+      rescue SyntaxError => e
+        Jekyll.logger.warn "YAML Exception reading #{filename}: #{e.message}"
+      # rubocop:disable RescueException
+      rescue Exception => e
```
You can probably change this to `rescue => e`.

parkr commented on lib/jekyll/convertible.rb:
```diff
@@ -296,5 +127,42 @@ def [](property)
+        raise(
+          Errors::InvalidYAMLFrontMatterError,
+          "Invalid YAML front matter in #{filename}"
+        )
```
I'd rather not use parentheses. How about:

```ruby
raise Errors::InvalidYAMLFrontMatterError,
  "Invalid YAML front matter in #{filename}"
```

parkr commented on lib/jekyll/document.rb:
```diff
@@ -103,7 +84,7 @@ def output_ext
     # Returns the basename without the file extname.
     def basename_without_ext
-      @basename_without_ext ||= File.basename(path, '.*')
+      @basename_without_ext ||= File.basename(path, File.extname(path))
```
This may change behaviour. Please revert.

parkr commented on lib/jekyll/document.rb:
```diff
@@ -237,18 +218,16 @@ def destination(base_directory)
-        f.write(output)
-      end
-
+      File.write(path, output)
```
Please ensure we are using the `wb` mode.

parkr commented on lib/jekyll/document.rb:
```diff
@@ -263,75 +242,20 @@ def read(opts = {})
         rescue SyntaxError => e
           Jekyll.logger.error "Error:", "YAML Exception reading #{path}: #{e.message}"
+        # rubocop: disable RescueException
         rescue Exception => e
```
You can remove `Exception` – just do `rescue => e` here.

ayastreb commented:
Hi @parkr, switching Pathname for Pathutil was proposed to me by @envygeeks in one of the previous reviews (#4960), so I thought it is safe :)
Should I revert it and open a separate PR for that?

parkr commented on lib/jekyll/document.rb:
```diff
@@ -433,15 +353,129 @@ def respond_to?(method, include_private = false)
+        Jekyll.logger.warn(
+          "Deprecation:",
+          "Document##{method} is now a key in the #data hash."
+        )
```
Let's do `Jekyll::Deprecator. deprecation_message "Document##{method} is now a key in the #data hash."`

parkr commented on lib/jekyll/document.rb:
```diff
@@ -433,15 +353,129 @@ def respond_to?(method, include_private = false)
+          data["date"].to_s,
+          "Document '#{relative_path}' does not have a valid date in the #{source}."
+        )
+      end
```
Why don't we do a `merge_date!` here, like `merge_categories!`.

parkr commented:
> Should I revert it and open a separate PR for that?

It appears to be working so we can leave it. But keeping PR's small makes them more manageable for us maintainers 😄

parkr commented on lib/jekyll/document.rb:
```diff
@@ -433,15 +353,129 @@ def respond_to?(method, include_private = false)
+          data,
+          "category",
+          "categories"
+        )
```
This looks like the wrong indentation. Would you push `Array` two spaces right and align the following lines?

```ruby
def populate_categories
  merge_data!({
    "categories" =>
      Array(data["categories"]) +
      Utils. pluralized_array_from_hash(data, "category", "categories")
  })
end
```

Does that work?

envygeeks replied:
Personally I'd format it as:

```ruby
def populate_categories
  merge_data!({
    "categories" =>
      Array(data["categories"]) + Utils. pluralized_array_from_hash(
        data, "category", "categories"
      )
  })
end

parkr commented on lib/jekyll/renderer.rb:
```diff
@@ -185,10 +219,21 @@ def converter_output_ext
+    private
+    def relative_path
+      document.relative_path
+    end
```
These two methods are superfluous – there is no reason to make a method here, and `Renderer#relative_path` doesn't make sense because the Renderer doesn't have any concept of path.

parkr commented on test/test_page.rb:
```diff
@@ -11,10 +11,10 @@ def setup_page(*args)
       "default" => Layout.new(@site, source_dir("_layouts"), "simple.html")
     }
-    page.render(layouts, @site.site_payload)
+    page.render(@site.site_payload)
```
We cannot change this method signature. Please revert.

parkr commented on lib/jekyll/renderer.rb:
```diff
@@ -13,170 +13,204 @@ def initialize(site, document, site_payload = nil)
+            "Conversion error:",
+            "#{converter.class} encountered an error "\
+            "while converting '#{relative_path}':"
+          )
```
Let's do an alternative route here:

```ruby
Jekyll.logger.error "Conversion error:",
  "#{converter.class} encountered an error while converting '#{document.relative_path}':"
```

parkr commented on lib/jekyll/renderer.rb:
```diff
@@ -13,170 +13,204 @@ def initialize(site, document, site_payload = nil)
+        document.output = render_liquid(document.content, path)
       end
 
-      Jekyll.logger.debug "Rendering Markup:", document.relative_path
```
What happened to this log message?

parkr commented:
@ayastreb I think at this time I would like you to split these out into multiple PR's. It's just too much all at once. Could you please take each file, and only do style-required refactors? Then refactors which change things more substantively could come later (e.g. Pathutil and changing functions around).

ayastreb commented:
@parkr thanks for the feedback!
I will try to do separate PRs today.
